### PR TITLE
Arrow,Core,Spark: Remove functionality marked for removal in 1.4.0

### PR DESCRIPTION
**Pull request by nastra**

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java [hunk 1]
```diff
@@ -46,6 +46,7 @@
  * RewriteDataFiles#TARGET_FILE_SIZE_BYTES}.
  *
  * @deprecated since 1.3.0, will be removed in 1.4.0; use {@link SizeBasedFileRewriter} instead.
+ *     Note: This can only be removed once Spark 3.1 + 3.2 isn't using this API anymore.
```
Comment (ConeyLiu) [inline core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:49]: Maybe we should update the Spark 3.1/3.2 usage in some other PRs.

**Comment by rdblue** [comment]
Looks like this includes changes from other PRs, so I'll hold off on reviewing until it's rebased. Just ping me when this is unblocked.

**Comment by nastra** [comment]
Sorry I forgot to mention it, but yes this currently depends on the changes from https://github.com/apache/iceberg/pull/8099

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java (hunk 1 above)
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:49]: yes that would be great, but I'm not sure how feasible this is, given that we don't backport every feature to old Spark versions

**Comment by nastra** [comment]
@rdblue this is ready for review now

File: core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java (hunk 1 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/actions/BinPackStrategy.java:49]: I think that would be a non-trivial process, which is probably not worth it.

**Review by aokolnychyi — APPROVED** [review]

**Comment by aokolnychyi** [comment]
Thanks, @nastra!

**Comment by nastra** [comment]
thanks for reviewing and merging this @aokolnychyi

**Pull request merged**